### PR TITLE
terminate envoy when number of active connections is zero

### PR DESCRIPTION
**Pull request by ramaraochavali**
For https://github.com/istio/istio/issues/34855 except that it still uses terminationDrainDuration.

It gets `downstream_cx_active` of all listeners (ignores admin and stats listeners) and when the total becomes zero, it terminates Envoy or waits for terminationDrainDuration. It starts this process after waiting for minDrainDuration i.e. 5s.

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@howardjohn @kyessenov I was experimenting this a bit and here is draft PR. Please take a look - if the approach makes sense - I will add tests.

File: pkg/config/mesh/mesh.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ func DefaultProxyConfig() meshconfig.ProxyConfig {
 		ServiceCluster:           constants.ServiceClusterName,
 		DrainDuration:            types.DurationProto(45 * time.Second),
 		ParentShutdownDuration:   types.DurationProto(60 * time.Second),
-		TerminationDrainDuration: types.DurationProto(5 * time.Second),
+		TerminationDrainDuration: types.DurationProto(45 * time.Second),
```
Comment (howardjohn) [inline pkg/config/mesh/mesh.go:46]: Discussed a bit more in the issue, but we need to make sure we don't make this a breaking API change

File: pkg/envoy/agent.go [hunk 2]
```diff
@@ -17,22 +17,43 @@ package envoy
...
 )
 
 var errAbort = errors.New("epoch aborted")
 
 const errOutOfMemory = "signal: killed"
 
+// minDrainDuration is the minimum duration for which agent waits before it actively checks
+// for outstanding connections. After this period, agent terminates proxy whenever active
+// connections become zero.
+var minDrainDuration = 5 * time.Second
```
Comment (howardjohn) [inline pkg/envoy/agent.go:37]: same comment about breaking change. You should be able to override to 0s

File: pkg/envoy/agent.go [hunk 3]
```diff
@@ -17,22 +17,43 @@ package envoy
...
+
+var activeConnectionCheckDelay = 1 * time.Second
+
+var knownIstioListeners = sets.NewSet(
+	"listener.0.0.0.0_15006.downstream_cx_active", "listener.0.0.0.0_15021.downstream_cx_active",
```
Comment (howardjohn) [inline pkg/envoy/agent.go:42]: The ports can be configured

File: pkg/envoy/agent.go [hunk 4]
```diff
@@ -105,13 +131,72 @@ func (a *Agent) terminate() {
 	if e != nil {
 		log.Warnf("Error in invoking drain listeners endpoint %v", e)
 	}
-	log.Infof("Graceful termination period is %v, starting...", a.terminationDrainDuration)
-	time.Sleep(a.terminationDrainDuration)
-	log.Infof("Graceful termination period complete, terminating remaining proxies.")
-	a.abortCh <- errAbort
+	time.Sleep(minDrainDuration)
```
Comment (howardjohn) [inline pkg/envoy/agent.go:134]: on L129 log what we are going to do maybe. Like `Agent draining proxy for 5s, then waiting for active connections to terminate...`

File: pkg/envoy/agent.go [hunk 5]
```diff
@@ -17,22 +17,43 @@ package envoy
...
+	"listener.admin.main_thread.downstream_cx_active",
+)
+
 // NewAgent creates a new proxy agent for the proxy start-up and clean-up functions.
-func NewAgent(proxy Proxy, terminationDrainDuration time.Duration) *Agent {
+func NewAgent(proxy Proxy, terminationDrainDuration time.Duration, localhost string, adminPort uint16) *Agent {
 	return &Agent{
 		proxy:                    proxy,
 		statusCh:                 make(chan exitStatus, 1), // context might stop drainage
+		drainCh:                  make(chan bool),
```
Comment (howardjohn) [inline pkg/envoy/agent.go:52]:
```suggestion
		drainCh:                  make(chan struct{}),
```

then use `close(drainCh)` to signal, not send

**Comment by kyessenov** [comment]
The cleanest solution is to wait for other containers to complete IMHO (and dealing with their organic restarts). 

What you have here seems reasonable. The danger is some app might want to open a connection at the end (think heavy app generating logs and dumping them somewhere else). So this won't work in general.

I'm actually would be totally OK if we stopped using envoy for readiness and prometheus ports. I don't like the complexity incurred by having to deal with those listeners. It's complexity for complexity sake.

**Comment by ramaraochavali** [comment]
@howardjohn @kyessenov Some how this branch is messedup. Closing this and opening a new PR https://github.com/istio/istio/pull/35059.

@howardjohn addressed all your comments. If we are good with the approach I will add tests.

**Comment by howardjohn** [comment]
> The cleanest solution is to wait for other containers to complete IMHO (and dealing with their organic restarts).

I agree but don't know how to do this. Any ideas?

**Pull request closed without merging**